jep/0000/README.adoc

```diff
@@ -0,0 +1,348 @@
+= JEP-0000: WebSocket Services
+:toc: preamble
+:toclevels: 3
+ifdef::env-github[]
+:tip-caption: :bulb:
+:note-caption: :information_source:
+:important-caption: :heavy_exclamation_mark:
+:caution-caption: :fire:
+:warning-caption: :warning:
+endif::[]
+
+.Metadata
+[cols="1h,1"]
+|===
+| JEP
+| 0000
+
+| Title
+| WebSocket Services
+
+| Sponsor
+| link:https://github.com/jglick[Jesse Glick]
+
+// Use the script `set-jep-status <jep-number> <status>` to update the status.
+| Status
+| Not Submitted :information_source:
+
+| Type
+| Standards
+
+| Created
+| 2019-11-21
+
+| BDFL-Delegate
+| TBD
+
+|===
+
+== Abstract
+
+Jenkins offers an API allowing WebSocket services to be implemented.
+This is available using the Stapler `HttpResponse` idiom,
+served as the final action after the regular Stapler routing mechanism.
+Unlike regular HTTP endpoints, bidirectional streaming communication is possible.
+
+The principal service exposed in this way is a new Remoting transport for inbound build agents at an HTTP(S) URL.
+The agent itself (typically named `agent.jar`) includes a WebSocket client capable of connecting to this endpoint.
+This provides an alternative to using the `JNLP4-connect` protocol on the Jenkins TCP port.
+Otherwise the regular behavior of inbound agents (`JNLPLauncher`) applies.
+
+The Jenkins CLI could also be served via WebSocket in place of the SSHD port or the problematic “full-duplex HTTP” transport.
+
+The implementation relies on APIs avalable in the built-in Jetty (“Winstone”) servlet container,
+but Jetty APIs are not reëxported to plugins.
```

Comment:
```suggestion
but Jetty APIs are not re-exported to plugins.
```

Reply:
Either is valid; I am using the older notation.

Reply:
True, but the newer form would be more recognizable and readable to more people.